File: scripts/forensic_control_test.mjs
import { readFileSync } from 'fs';
import { resolve } from 'path';

console.log('🧪 ==================================================================');
console.log('🧪 FORENSIC CONTROL TEST: UNTRAINED vs MATHEMATICAL INTERPOLATION');
console.log('🧪 ==================================================================\n');

async function main() {
  const ort = await import('onnxruntime-web');

  const buf2 = readFileSync(resolve(process.cwd(), 'public/models/espcn-x2.onnx'));
  const buf4 = readFileSync(resolve(process.cwd(), 'public/models/espcn-x4.onnx'));

  const session2 = await ort.InferenceSession.create(buf2);
  const session4 = await ort.InferenceSession.create(buf4);

  // Test Pattern 1: Flat gray (0.5)
  const flat = new Float32Array(256 * 256).fill(0.5);

  // Test Pattern 2: Step edge
  const edge = new Float32Array(256 * 256);
  for (let y = 0; y < 256; y++) {
    for (let x = 0; x < 256; x++) {
      edge[y * 256 + x] = x >= 128 ? 1.0 : 0.0;
    }
  }

  // Test Pattern 3: Checkerboard
  const checker = new Float32Array(256 * 256);
  for (let y = 0; y < 256; y++) {
    for (let x = 0; x < 256; x++) {
      checker[y * 256 + x] = ((Math.floor(x / 8) + Math.floor(y / 8)) % 2 === 0) ? 1.0 : 0.0;
    }
  }

  async function evaluate(name, data) {
    console.log(`--- Test Pattern: ${name} ---`);
    const inTensor = new ort.Tensor('float32', data, [1, 1, 256, 256]);

    // 2x
    const res2 = await session2.run({ input: inTensor });
    const out2 = res2.output.data;
    let sum2 = 0;
    let min2 = Infinity;
    let max2 = -Infinity;
    for (let i = 0; i < out2.length; i++) {
      sum2 += out2[i];
      if (out2[i] < min2) min2 = out2[i];
      if (out2[i] > max2) max2 = out2[i];
    }
    const mean2 = sum2 / out2.length;

    console.log(`  ESPCN 2× Output:`);
    console.log(`    Mean: ${mean2.toFixed(4)}, Min: ${min2.toFixed(4)}, Max: ${max2.toFixed(4)}`);

    // 4x
    const res4 = await session4.run({ input: inTensor });
    const out4 = res4.output.data;
    let sum4 = 0;
    let min4 = Infinity;
    let max4 = -Infinity;
    for (let i = 0; i < out4.length; i++) {
      sum4 += out4[i];
      if (out4[i] < min4) min4 = out4[i];
      if (out4[i] > max4) max4 = out4[i];
    }
    const mean4 = sum4 / out4.length;

    console.log(`  ESPCN 4× Output:`);
    console.log(`    Mean: ${mean4.toFixed(4)}, Min: ${min4.toFixed(4)}, Max: ${max4.toFixed(4)}\n`);
  }

  await evaluate('Flat Color (0.500)', flat);
  await evaluate('Sharp Step Edge (0.0 / 1.0)', edge);
  await evaluate('Checkerboard Pattern (0.0 / 1.0)', checker);
}

main().catch(console.error);

File: scripts/forensic_control_test.py
import sys
import numpy as np
import onnxruntime as ort

if sys.platform == 'win32':
    sys.stdout.reconfigure(encoding='utf-8')

print("🧪 ==================================================================")
print("🧪 FORENSIC CONTROL TEST: UNTRAINED vs MATHEMATICAL INTERPOLATION")
print("🧪 ==================================================================\n")

# Test Patterns
# Pattern 1: Flat gray (0.5)
flat_input = np.full((1, 1, 256, 256), 0.5, dtype=np.float32)

# Pattern 2: High contrast sharp step edge
edge_input = np.zeros((1, 1, 256, 256), dtype=np.float32)
edge_input[:, :, :, 128:] = 1.0

# Pattern 3: High frequency checkerboard
checker_input = np.zeros((1, 1, 256, 256), dtype=np.float32)
for y in range(256):
    for x in range(256):
        if (x // 8 + y // 8) % 2 == 0:
            checker_input[0, 0, y, x] = 1.0

session2 = ort.InferenceSession('public/models/espcn-x2.onnx')
session4 = ort.InferenceSession('public/models/espcn-x4.onnx')

def evaluate_pattern(name, inp_data):
    print(f"--- Test Pattern: {name} ---")
    
    # 2x Inference
    res2 = session2.run(None, {'input': inp_data})[0]
    mean_in = np.mean(inp_data)
    mean_out2 = np.mean(res2)
    std_out2 = np.std(res2)
    min_out2 = np.min(res2)
    max_out2 = np.max(res2)
    
    print(f"ESPCN 2× Output:")
    print(f"  Input Mean: {mean_in:.4f} -> Output Mean: {mean_out2:.4f}")
    print(f"  Output Min: {min_out2:.4f}, Max: {max_out2:.4f}, Std: {std_out2:.4f}")
    
    # 4x Inference
    res4 = session4.run(None, {'input': inp_data})[0]
    mean_out4 = np.mean(res4)
    std_out4 = np.std(res4)
    min_out4 = np.min(res4)
    max_out4 = np.max(res4)
    
    print(f"ESPCN 4× Output:")
    print(f"  Input Mean: {mean_in:.4f} -> Output Mean: {mean_out4:.4f}")
    print(f"  Output Min: {min_out4:.4f}, Max: {max_out4:.4f}, Std: {std_out4:.4f}")
    print("")

evaluate_pattern("Flat Color (0.500)", flat_input)
evaluate_pattern("Sharp Step Edge", edge_input)
evaluate_pattern("High-Frequency Checkerboard", checker_input)
